app/lib/shared/task_scheduler.dart

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:async/async.dart';
+import 'package:logging/logging.dart';
+
+final Logger _logger = new Logger('pub.scheduler');
+
+/// Interface for task execution.
+typedef Future TaskRunner(Task task);
+
+// ignore: one_member_abstracts
+abstract class TaskSource {
+  /// Returns a stream of currently available tasks at the time of the call.
+  Stream<Task> currentTasks();
+}
+
+/// Tasks coming from through the isolate's receivePort, originating from a
+/// HTTP handler that received a ping after a new upload.
+class ManualTriggerTaskSource implements TaskSource {
+  final Stream taskReceivePort;
+  final int capacity;
+  StreamSubscription _subscription;
+  final Set<Task> _triggered = new Set();
+
+  ManualTriggerTaskSource(this.taskReceivePort, {this.capacity: 100}) {
+    _subscription = taskReceivePort.listen((Task task) {
+      // protect against spamming the manual trigger
+      if (_triggered.length < capacity) {
+        _triggered.add(task);
+      }
```

Comment:
Consider logging if we ever hit more than 100 manual tasks, since we never expect it.

You can do it in an upcoming CL.